delivery_correos_express/models/delivery_carrier.py

```diff
@@ -0,0 +1,260 @@
+        picking.tracking_state_history = "\n".join(
+            [
+                "{} {} - [{}] {}".format(
+                    "{}:{}:{}".format(
+                        t.get("horaEstado")[:2],
+                        t.get("horaEstado")[2:-2],
+                        t.get("horaEstado")[-2:],
+                    ),
+                    "{}/{}/{}".format(
+                        t.get("fechaEstado")[:2],
+                        t.get("fechaEstado")[2:-4],
+                        t.get("fechaEstado")[4:],
+                    ),
+                    t.get("codEstado"),
+                    t.get("descEstado"),
+                )
+                for t in tracking_events
+            ]
+        )
+        tracking = tracking_events.pop()
+        picking.tracking_state = "[{}] {}".format(
+            tracking.get("codEstado"), tracking.get("descEstado")
+        )
+
+    def correos_express_cancel_shipment(self, pickings):
+        for picking in pickings.filtered("carrier_tracking_ref"):
+            picking.message_post(
+                body=_(
+                    "Correos Express does not provide a method to cancel a shipment "
+                    "that has been registered. If you need to change some information "
+                    "you, create a new shipment with a new label. This doesn't mean "
+                    "that the shipment will be invoiced, this only happens if the "
+                    "package is picked up and it enters the shipping stage"
+                )
+            )
+
+    def _prepare_print_label(self, carrier_tracking_ref):
+        return {
+            "nenvio": carrier_tracking_ref,
+            "tipo": self.correos_express_label_type,
+            "keyCli": self.correos_express_customer_code,
+        }
+
+    def correos_express_get_label(self, carrier_tracking_ref):
+        self.ensure_one()
+        if not carrier_tracking_ref:
+            return False
+        correos_express_request = CorreosExpressRequest(self)
+        labels = correos_express_request.print_shipment(
+            self._prepare_print_label(carrier_tracking_ref)
+        )
+        return labels or False
+
+    def correos_express_rate_shipment(self, order):
+        """Not implemented with Correos, these values are so it works with websites"""
+        return {
+            "success": True,
+            "price": self.product_id.lst_price,
+            "error_message": _(
+                "Correos Express API doesn't provide methods to compute "
+                "delivery rates, so you should rely on another price method "
+                "instead or override this one in your custom code."
+            ),
+            "warning_message": _(
+                "Correos Express API doesn't provide methods to compute "
+                "delivery rates, so you should rely on another price method "
+                "instead or override this one in your custom code."
+            ),
+        }
```

Comment:
```suggestion
      raise NotImplementedError(
            _(
                """
                Correos Express API doesn't provide methods to compute delivery rates, so
            you should relay on another price method instead or override this
            one in your custom code.
        """
            )
        )
```

Reply:
This returns a dict in order to not fail when using this from a website interface, with the dict it doesnt fail but warns you it might not be accurate. I don't like the idea of it simply not being usable from websites without custom code, I'd rather it work and the user can change the price if need be afterwards

Reply:
I think it is necessary if the carrier is configured this way
![odoo-carrier](https://user-images.githubusercontent.com/4117568/140897756-e9770ed8-109d-4b0d-a414-4bfd671d60b6.png)


But it wouldn't be necessary if another pricing method is used, what do you think?

Reply:
Where is that config? I don't see it; Either way I still prefer having it not error and instead just warn

Reply:
Sorry, this is related to the `delivery_price_method` addon (which is not migrated to v14 yet).
It is not necessary then to make any changes.